### PR TITLE
New TF Serving template

### PR DESCRIPTION
**Pull request by lluunn**
For #1264

Move things into jsonnet instead of relying on import.
It's easier to customize.

- previous params remain the same, just the component name is tf-serving-template
- e2e test is now using the new tf-serving-template

/cc @jlewi

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kubeflow/1589)

### CODE REVIEW

**Comment by jlewi** [comment]
I think we're headed in the direction of having multiple prototypes (we already have 2) this adds a 3rd.
I think if we follow my original suggestion of putting everything into the prototype (.jsonnet) file we will get a lot of unnecessary code duplication that will be difficult to maintain overtime.

@kkasravi's recent prototype #1501 suggests to me (see [comment](https://github.com/kubeflow/kubeflow/issues/376#issuecomment-422377216)) that we should define a base libsonnet file and then use k8s libsonnet to modify it for each prototype.

I considered having a basic prototype with all the logic in the .jsonnet file that people can use as a simple example so they don't have to learn the k8s lib. But even that seems like it would get pretty unruly.

My initial thought is that we should have the following prototypes 

* TFServing using S3
* TFServing using GCS
* TFServing using PVC

So just supporting those 3 different use cases will lead to a lot of code duplication if we put all the code into the .jsonnet file.

**Comment by lluunn** [comment]
I was thinking the we can use only this one jsonnet later.
The optional features are not mutually exclusive. They (GCP credential, S3, istio, request log) can co-exist, so have their own "enabled" flag.

Here the jsonnet structure has two parts (before and after line 75):
1. Flags and customization.
    Flags for enabling features and can be easily set. Customization for each feature can be set by flags 
    or by modify the component directly (the component is the copy of whole jsonnet)
1. Service/Container/Deployment
    This part may look a little complicated as features being added, `env: [] + if XX then [a] else [] + ...`
    But it should not need to be changed by users.

WDYT?

**Comment by jlewi** [comment]
One of the reasons for having parameters is for making it easy for users to figure out what they have to set.
The problem with the way its currently configured is that the parameters don't define the things the user has to set for S3 vs. GCS. They'd have to read the code to realize there are parameters like "gcpCredentialSecretName" and "s3SecretName".

If we have separate prototypes for GCP and S3. Then in the prototype for GCP we can explicitly define all the relevant parameters e.g.

```
// @optionalParam gcpSecretName string "user-sa" Name of the secret containing the GCP service account
```

Then when the user invokes help on the component they can get a meaningful list of parameters that they could set.

**Comment by lluunn** [comment]
But how about istio and request logs? They can co-exist with gcp credentials and S3.
We can't make one prototype just for istio and another just for request logs because users might want both (and gcp credential as well).

We can add both gcpSecretName and s3Enabled as flags in this prototype, and add better documentation. WDYT?

File: kubeflow/tf-serving/prototypes/tf-serving.jsonnet [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+        port: 8000,
+        targetPort: 8000,
+      },
+    ],
+    selector: {
+      app: appName,
+    },
+    type: "ClusterIP",
+  },
+};  // service
```
Comment (kkasravi) [inline kubeflow/tf-serving/prototypes/tf-serving.jsonnet]: I would add a 
service:: service

as described [here](https://github.com/databricks/jsonnet-style-guide#defining-classes)

File: kubeflow/tf-serving/prototypes/tf-serving.jsonnet [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+  image: modelServerImage,
+  imagePullPolicy: "IfNotPresent",
+  name: name,
+  ports: [
+    {
+      containerPort: 9000,
+    },
+  ],
+  env: []
+       + if util.toBool(params.s3Enable) then s3Env else []
```
Comment (kkasravi) [inline kubeflow/tf-serving/prototypes/tf-serving.jsonnet]: I would move this into deployment.mapContainers  in the kubeflow/tf-serving/prototypes/tf-serving-aws.jsonnet file as described above

File: kubeflow/tf-serving/prototypes/tf-serving.jsonnet [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+        },
+      },
+      spec: {
+        containers: [
+          modelServerContainer,
+        ] + if util.toBool(params.deployHttpProxy) then [
+          httpProxyContainer,
+        ] else [],
+        volumes: []
+                 + if gcpParams.gcpCredentialSecretName != "" then
```
Comment (kkasravi) [inline kubeflow/tf-serving/prototypes/tf-serving.jsonnet]: move into a kubeflow/tf-serving/prototypes/tf-serving-gcp.jsonnet file

**Comment by lluunn** [comment]
Thanks @jlewi @kkasravi , PTAL

**Comment by lluunn** [comment]
argo test failed in kfctl
Failed to run argo workflow

**Comment by lluunn** [comment]
argo test failed again, there might be an issue
http://testing-argo.kubeflow.org/workflows/kubeflow-test-infra/kubeflow-presubmit-kfctl-1589-636f49f-3514-3e66?tab=workflow

**Comment by lluunn** [comment]
still argo test failure, but I don't see how this PR is related to that.

**Comment by lluunn** [comment]
https://pantheon.corp.google.com/logs/viewer?project=kubeflow-ci&minLogLevel=0&expandAll=false&timestamp=2018-09-26T20:45:52.695000000Z&customFacets=&limitCustomFacetWidth=true&interval=NO_LIMIT&resource=container%2Fcluster_name%2Fkubeflow-testing&scrollTimestamp=2018-09-26T18:07:57.000000000Z&advancedFilter=resource.type%3D%22container%22%0Aresource.labels.cluster_name%3D%22kfctl-7247%22%0Aresource.labels.pod_id%3D%22hello-world-p822r%22%0A

I can see the argo pod up at 11:07:47, but it timeout at 07:03. So I am extending the wait time.

File: kubeflow/tf-serving/util.libsonnet [hunk 4]
```diff
@@ -1,5 +1,7 @@
 // Some useful routines.
 {
+  local k = import "k.libsonnet",
+
   // Convert a string to upper case.
   upper:: function(x) {
```
Comment (kkasravi) [inline kubeflow/tf-serving/util.libsonnet]: std library has std.asciiUpper(str)
http://jsonnet.org/ref/stdlib.html

**Comment by kkasravi** [comment]
@lluunn jfmt errors

**Comment by lluunn** [comment]
Thanks @kkasravi

**Comment by jlewi** [comment]
This looks great.
/lgtm
/approve

**Comment by lluunn** [comment]
not sure why notebook release workflow is triggered...

**Pull request merged**